ostd/src/mm/frame/segment.rs

```diff
@@ -1,178 +1,271 @@
 // SPDX-License-Identifier: MPL-2.0
 
-//! A contiguous segment of untyped memory pages.
+//! A contiguous range of frames.
 
-use core::ops::Range;
+use core::{mem::ManuallyDrop, ops::Range};
 
-use crate::{
-    mm::{
-        frame::FrameMeta,
-        io::{FallibleVmRead, FallibleVmWrite},
-        page::ContPages,
-        Frame, HasPaddr, Infallible, Paddr, VmIo, VmReader, VmWriter,
-    },
-    Error, Result,
-};
+use super::{inc_frame_ref_count, meta::FrameMeta, Frame};
+use crate::mm::{Paddr, UFrameMeta, PAGE_SIZE};
 
-/// A contiguous segment of untyped memory pages.
+/// A contiguous range of homogeneous physical memory frames.
 ///
-/// A [`Segment`] object is a handle to a contiguous range of untyped memory
-/// pages, and the underlying pages can be shared among multiple threads.
-/// [`Segment::slice`] can be used to clone a slice of the segment (also can be
-/// used to clone the entire range). Reference counts are maintained for each
-/// page in the segment. So cloning the handle may not be cheap as it
-/// increments the reference count of all the cloned pages.
+/// This is a handle to many contiguous frames. It will be more lightweight
+/// than owning an array of frame handles.
 ///
-/// Other [`Frame`] handles can also refer to the pages in the segment. And
-/// the segment can be iterated over to get all the frames in it.
+/// The ownership is achieved by the reference counting mechanism of frames.
+/// When constructing a [`Segment`], the frame handles are created then
+/// forgotten, leaving the reference count. When dropping a it, the frame
+/// handles are restored and dropped, decrementing the reference count.
 ///
-/// To allocate a segment, use [`crate::mm::FrameAllocator`].
-///
-/// # Example
-///
-/// ```rust
-/// let vm_segment = FrameAllocOptions::new(2)
-///     .is_contiguous(true)
-///     .alloc_contiguous()?;
-/// vm_segment.write_bytes(0, buf)?;
-/// ```
+/// All the metadata of the frames are homogeneous, i.e., they are of the same
+/// type.
 #[derive(Debug)]
-pub struct Segment {
-    pages: ContPages<FrameMeta>,
+#[repr(transparent)]
+pub struct Segment<M: FrameMeta + ?Sized> {
+    range: Range<Paddr>,
+    _marker: core::marker::PhantomData<M>,
 }
 
-impl HasPaddr for Segment {
-    fn paddr(&self) -> Paddr {
-        self.pages.start_paddr()
+/// A contiguous range of homogeneous physical memory frames that have any metadata.
+///
+/// In other words, the metadata of the frames are of the same type but the type
+/// is not known at compile time. An [`DynSegment`] as a parameter accepts any
+/// type of segments.
+///
+/// The usage of this frame will not be changed while this object is alive.
+pub type DynSegment = Segment<dyn FrameMeta>;
```

Comment:
No one seems to be using this trait alias except the `segment` module itself. So may be this is not very useful any way.